fedimint-atomic-broadcast/src/db.rs

```diff
@@ -0,0 +1,133 @@
+use std::io::Cursor;
+
+use fedimint_core::db::Database;
+use fedimint_core::encoding::{Decodable, Encodable};
+use fedimint_core::impl_db_record;
+
+use crate::SignedBlock;
+
+#[derive(Debug, Encodable, Decodable)]
+struct SignedBlockKey(u64);
+
+impl_db_record!(
+    key = SignedBlockKey,
+    value = SignedBlock,
+    db_prefix = 0x00,
+    notify_on_modify = false,
+);
+
+#[derive(Debug, Encodable, Decodable)]
+struct AlephBufferKey(u64, u64);
+
+impl_db_record!(
+    key = AlephBufferKey,
+    value = Vec<u8>,
+    db_prefix = 0x01,
+    notify_on_modify = false,
+);
+
+pub async fn load_block(db: &Database, index: u64) -> Option<SignedBlock> {
+    db.begin_transaction()
+        .await
+        .get_value(&SignedBlockKey(index))
+        .await
+}
+
+pub async fn open_session(db: Database, index: u64) -> (Cursor<Vec<u8>>, UnitSaver) {
+    let mut buffer = vec![];
+    let mut buffer_index = 0;
+    let mut transaction = db.begin_transaction().await;
+
+    while let Some(bytes) = transaction
+        .get_value(&AlephBufferKey(index, buffer_index))
+        .await
+    {
+        buffer.extend(bytes);
+        buffer_index += 1;
+    }
+
+    std::mem::drop(transaction);
+
+    tracing::info!("Loaded aleph buffer with {} bytes", buffer.len());
+
+    // the cursor enables aleph bft to read all previously flushed bytes via
+    // std::io::Read
+    let loader = Cursor::new(buffer);
+
+    // aleph bft expects bytes written via the saver to be appended to the bytes
+    // previously read via the loader - hence we pass the first free buffer
+    // index to the UnitSaver as an offset
+    let saver = UnitSaver::new(db, index, buffer_index);
+
+    (loader, saver)
+}
+
+pub struct UnitSaver {
+    db: Database,
+    index: u64,
```

Comment:
It might be because I don't understand the context, but `index`, `buffer` and `buffer_index` don't seem like great names. When there's an index what's interesting is what is an index of, similar with "buffer". "buffer_index" between a "index" and "buffer" is very confusing. :D

Reply:
I have improved the naming and added some comments.